### ISSUE TITLE
All or nothing yet again :(

### ISSUE THREAD

**Issue by forkoz** [SilenceIM/Silence#415]
I've tried quite a few encrypted SMS programs, why does every single one of them demand to be king of the hill?

Silence IM would not receive messages along with my standard SMS client even when given full permissions. I had to set it as the default client or import my 50k sms mesages into its DB.

I really don't like this approach, nor the material interface. My 4.4 messaging is nice for everyone I don't wish to encrypt communication with. At least you aren't starting on boot over and over like signal and kontalk and restoring the permissions.

Is it so bad to want a separate encrypted SMS client or at least to have that option and not mix it with my plaintext?

**Comment by ArchangeGabriel** [comment]
I’m not sure about this, but when did Google introduced the fact that one and only one app could be registered as the default SMS app and thus receive and send them? Because what you’re telling us about here is Android normal behaviour since then, and I think it was exactly in 4.4 that this was introduced.

In which case there is nothing that can be done here.

**Comment by pR0Ps** [comment]
Google changed the way SMS applications work in KitKat (v4.4).

In Jelly Bean and lower, app developers could set an SMS priority level of sorts. The app with the highest priority would see the incoming SMS, act on it, then decide if it wanted to pass it to the next app in line. This made it easy at the time - we bumped the priority to always be first, checked the message, and only passed on unencrypted messages to the other apps.

In KitKat+ Android now sends 2 broadcasts when receiving an SMS:
 - The `SMS_RECEIVED_ACTION` intent is sent to every app.
 - The `SMS_DELIVER_ACTION` intent is only sent to the default SMS app.

Therefore, in KitKat+ if you don't set Silence to be the default SMS app the following will happen:
 - All incoming SMS messages (encrypted and non-encrypted) will show up in whatever app is set as the default messenger (via `SMS_DELIVER_ACTION`).
 - All incoming encrypted SMS messages will be seen by Silence (via `SMS_RECEIVED_ACTION`), decrypted, and displayed in Silence. Silence will ignore non-encrypted messages and let the default app handle them.

This works, but there are two things about it that suck:
 - Your default SMS app will store and display large amounts of unreadable text for incoming encrypted messages. This is because we can't prevent the default SMS application from seeing the incoming SMS so it behaves normally and displays it like any other message.
 - You will receive 2 notifications for every encrypted SMS message you receive. One from Silence to let you know it received an encrypted message and one from your default SMS app.

**TL;DR:** It's not "All or nothing yet again", it's "All or a terrible experience (which you are free to choose)" and there's nothing we can do about it.

More details: http://android-developers.blogspot.ca/2013/10/getting-your-sms-apps-ready-for-kitkat.html
